process/block/sovereignChainBlock.go

```diff
@@ -0,0 +1,1279 @@
+	miniBlockHashes := make(map[int][]byte, len(miniBlockHeaders))
+	for i := 0; i < len(miniBlockHeaders); i++ {
+		miniBlockHashes[i] = miniBlockHeaders[i].GetHash()
+	}
+
+	log.Trace("cross mini blocks in body",
+		"num miniblocks", len(miniBlockHashes),
+	)
+
+	processedExtendedShardHeaders, err := scbp.getOrderedProcessedExtendedShardHeadersFromMiniBlockHashes(miniBlockHeaders, miniBlockHashes)
+	if err != nil {
+		return nil, err
+	}
+
+	return processedExtendedShardHeaders, nil
+}
+
+func (scbp *sovereignChainBlockProcessor) getOrderedProcessedExtendedShardHeadersFromMiniBlockHashes(
+	miniBlockHeaders []data.MiniBlockHeaderHandler,
+	miniBlockHashes map[int][]byte,
+) ([]data.HeaderHandler, error) {
+
+	processedExtendedShardHeaders := make([]data.HeaderHandler, 0, len(scbp.hdrsForCurrBlock.hdrHashAndInfo))
+	processedCrossMiniBlocksHashes := make(map[string]bool, len(scbp.hdrsForCurrBlock.hdrHashAndInfo))
+
+	scbp.hdrsForCurrBlock.mutHdrsForBlock.RLock()
+	for extendedShardHeaderHash, headerInfo := range scbp.hdrsForCurrBlock.hdrHashAndInfo {
+		if !headerInfo.usedInBlock {
+			continue
+		}
+
+		extendedShardHeader, ok := headerInfo.hdr.(*block.ShardHeaderExtended)
+		if !ok {
+			scbp.hdrsForCurrBlock.mutHdrsForBlock.RUnlock()
+			return nil, process.ErrWrongTypeAssertion
+		}
+
+		log.Trace("extended shard header",
+			"nonce", extendedShardHeader.GetNonce(),
+		)
+
+		crossMiniBlockHashes := make(map[string]struct{})
+		incomingMiniBlocks := extendedShardHeader.GetIncomingMiniBlocks()
+		for _, mb := range incomingMiniBlocks {
+			mbHash, err := core.CalculateHash(scbp.marshalizer, scbp.hasher, mb)
+			if err != nil {
+				scbp.hdrsForCurrBlock.mutHdrsForBlock.RUnlock()
+				return nil, err
+			}
+
+			crossMiniBlockHashes[string(mbHash)] = struct{}{}
+		}
+
+		for hash := range crossMiniBlockHashes {
+			processedCrossMiniBlocksHashes[hash] = scbp.processedMiniBlocksTracker.IsMiniBlockFullyProcessed([]byte(extendedShardHeaderHash), []byte(hash))
+		}
+
+		for key, miniBlockHash := range miniBlockHashes {
+			_, ok = crossMiniBlockHashes[string(miniBlockHash)]
+			if !ok {
+				continue
+			}
+
+			processedCrossMiniBlocksHashes[string(miniBlockHash)] = miniBlockHeaders[key].IsFinal()
+
+			delete(miniBlockHashes, key)
+		}
+
+		log.Trace("cross mini blocks in extended shard header",
+			"num miniblocks", len(crossMiniBlockHashes),
+		)
+
+		processedAll := true
+		for hash := range crossMiniBlockHashes {
+			if !processedCrossMiniBlocksHashes[hash] {
+				processedAll = false
+				break
+			}
+		}
+
+		if processedAll {
+			processedExtendedShardHeaders = append(processedExtendedShardHeaders, extendedShardHeader)
+		}
+	}
+	scbp.hdrsForCurrBlock.mutHdrsForBlock.RUnlock()
+
+	process.SortHeadersByNonce(processedExtendedShardHeaders)
+
+	return processedExtendedShardHeaders, nil
+}
+
+func (scbp *sovereignChainBlockProcessor) addProcessedCrossMiniBlocksFromExtendedShardHeader(headerHandler data.HeaderHandler) error {
```

Comment:
again, function too long

Reply:
This func is only 60 lines of code

Reply:
I have structs less than 60 lines of code